src/mumble/AudioInput.cpp

```diff
@@ -459,43 +548,20 @@ void AudioInput::addMic(const void *data, unsigned int nsamp) {
 				speex_resampler_process_float(srsMic, 0, pfMicInput, &inlen, pfOutput, &outlen);
 			}
 
+            // If echo cancellation is enabled the pointer ends up in the resynchronizer queue
+            // and may need to outlive this function's frame
+            short *psMic = iEchoChannels > 0 ? new short[iFrameSize] : (short*)alloca(iFrameSize*sizeof(short));
+            
 			// Convert float to 16bit PCM
 			const float mul = 32768.f;
 			for (int j = 0; j < iFrameSize; ++j)
 				psMic[j] = static_cast<short>(qBound(-32768.f, (ptr[j] * mul), 32767.f));
 
-			// If we have echo chancellation enabled...
-			if (iEchoChannels > 0) {
-				short *echo = NULL;
-
-				{
-					QMutexLocker l(&qmEcho);
-
-					if (qlEchoFrames.isEmpty()) {
-						iJitterSeq = 0;
-						iMinBuffered = 1000;
-					} else {
-						// Compensate for drift between the microphone and the echo source
-						iMinBuffered = qMin(iMinBuffered, qlEchoFrames.count());
-
-						if ((iJitterSeq > 100) && (iMinBuffered > 1)) {
-							iJitterSeq = 0;
-							iMinBuffered = 1000;
-							delete [] qlEchoFrames.takeFirst();
-						}
-						echo = qlEchoFrames.takeFirst();
-					}
-				}
-
-				if (echo) {
-					// We have echo data for the current frame, remember that
-					delete [] psSpeaker;
-					psSpeaker = echo;
-				}
-			}
-
-			// Encode and send frame
-			encodeAudioFrame();
+			// If we have echo cancellation enabled...
+			if(iEchoChannels > 0)
+			{
+				resync.addMic(psMic);
+			} else encodeAudioFrame(AudioChunk(psMic));
```

Comment:
I'd prefer a "proper" `else` block for this
```suggestion
			} else {
				encodeAudioFrame(AudioChunk(psMic));
			}
```